server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -257,11 +257,24 @@ public void run() {
           if (state == TabletState.ASSIGNED) {
             goal = TabletGoalState.HOSTED;
           }
+          if (Manager.log.isTraceEnabled()) {
+            Manager.log.trace(
+                "[{}] Shutting down all Tservers: {}, dependentCount: {} Extent: {}, state: {}, goal: {}",
+                store.name(), manager.serversToShutdown.equals(currentTServers.keySet()),
+                dependentWatcher == null ? "null" : dependentWatcher.assignedOrHosted(), tls.extent,
+                state, goal);
+          }
 
           // if we are shutting down all the tabletservers, we have to do it in order
           if ((goal == TabletGoalState.SUSPENDED && state == TabletState.HOSTED)
               && manager.serversToShutdown.equals(currentTServers.keySet())) {
-            if (dependentWatcher != null && dependentWatcher.assignedOrHosted() > 0) {
+            // If the stats object in the dependentWatcher is empty, then no tablets
+            // are being reported as needing to be hosted (everything is fine). If
+            // everything is fine, or there are still assigned or hosted tablets,
+            // then we cannot suspend the tablets for this watcher yet. Continue
+            // to host these tablets.
+            if (dependentWatcher != null && (dependentWatcher.getStats().isEmpty()
+                || dependentWatcher.assignedOrHosted() > 0)) {
```

Comment:
It would be less racy if both isEmpty and assignedOrHosted operated on the same stats map.  Being called at different times they could get different maps.  Could make the assignedOrHosted method a static util method that takes a map as an argument.

```suggestion
            // to host these tablets.
            var stats = dependentWatcher.getStats();
            if (dependentWatcher != null && (stats.isEmpty()
                || assignedOrHosted(stats) > 0)) {
```

Reply:
I just noticed the code suggestion I made could cause an NPE

Reply:
If dependentWatcher is null, then the following in the suggestion I made will blow up.

```java
var stats = dependentWatcher.getStats();
```

Reply:
Modified to use the same stats object in c0b9f14